blog/2021/01/snoopi_deep.md

```diff
@@ -0,0 +1,221 @@
+Starting with Julia 1.2, it became possible to do a limited amount of "spying" on inference, and this infrastructure corresponds to SnoopCompile's `@snoopi` macro.
+Julia 1.6 includes new changes that have permitted a far deeper look at what inference is doing.
+Appropriately enough, SnoopCompile calls this `@snoopi_deep`.
+
+The rich data collected by `@snoopi_deep` are useful for several different purposes.
+in this post, we'll describe the basic tool and show how it can be used to profile inference.
+Later posts will show other ways to use the data to reduce the amount of type-inference or cache its results.
+
+## Collecting the data
+
+To see `@snoopi_deep` in action, we'll use the following demo:
+
+```
+module SnoopDemo
+    struct MyType{T} x::T end
+
+    extract(y::MyType) = y.x
+
+    function domath(x)
+        y = x + x
+        return y*x + 2*x + 5
+    end
+
+    dostuff(y) = domath(extract(y))
+
+    function packintype(x)
+        y = MyType(x)
+        return dostuff(y)
+    end
+end
+```
+
+The main call, `packintype`, stores the input in a `struct`, and then calls functions that extract the field value and performs arithmetic on the result.
+To profile inference on this call, we simply do the following:
+
+```
+julia> using SnoopCompile
+
+julia> tinf = @snoopi_deep SnoopDemo.packintype(1)
+InferenceTimingNode: 0.009382/0.010515 on InferenceFrameInfo for Core.Compiler.Timings.ROOT() with 1 direct children
+```
+
+@@note
+**Box 1** Inference gets called only on the *first* invocation of a method with those specific types. You have to redefine the `SnoopDemo` module (by re-executing the command we used to define it) if you want to collect data with `@snoopi_deep` on the same code a second time.  This won't redefine any `Base` methods that `SnoopDemo` depends on, so you still may see differences between running this in a fresh session and subsequent uses of `@snoopi_deep`.
+@@
+
+This may not look like much, but there's a wealth of information hidden inside `tinf`.
+While optional, before proceeding further it's recommended to check `tinf` for hints that [invalidation](https://julialang.org/blog/2020/08/invalidations/) may have influenced the result:
+
+```
+julia> staleinstances(tinf)
+SnoopCompileCore.InferenceTiming[]
+```
+
+`staleinstances` extracts `MethodInstances` that have some "stale" (no longer callable) generated code.
+In our case, it returns an empty list, meaning that it found no stale instances, which guarantees that no invalidation occurred.
+There's nothing "funny" going on behind the scenes that will influence our results.
+
+@@note
+**Box 2** Unlike `@snoopr`, another macro exported by SnoopCompile, `staleinstances` does not distinguish between whether that stale code existed before you ran the `@snoopi_deep` block, or whether it got invalidated during the running of that block.
+For detailed analysis of invalidations, `@snoopr` is the recommended tool.
+@@
```

Comment:
🔥 this is super cool! You've made these changes after I stopped paying attention. thanks for all of this; it's neat! :)

Reply:
😂 I see Nathan and I had similar sentiments!